Add a Gaussian Splat and octomap rendering sample

gpx1000 opened the pull request:
## Description

This is a sample showing a complex interaction with other libraries to demonstrate how to work with libraries and tools int he wild.  Here Octomap is used to demonstrate Occupancy grid rendering. We also demonstrate how to use ImGUI in a project with added functionality and positioning of rendered assets.

## General Checklist:

Please ensure the following points are checked:

- [ ] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [ ] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [ ] I have commented any added functions (in line with Doxygen)
- [ ] I have commented any code that could be hard to understand
- [ ] My changes do not add any new compiler warnings
- [ ] My changes do not add any new validation layer errors or warnings
- [ ] I have used existing framework/helper functions where possible
- [ ] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [ ] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [ ] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

 If this PR contains framework changes:
 - [ ] I did a full batch run using the `batch` command line argument to make sure all samples still work properly

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [ ] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)
- [ ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [ ] For new samples, I have added a paragraph with a summary to the appropriate chapter in the readme of the folder that the sample belongs to [e.g. api samples readme](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/samples/api/README.adoc)
- [ ] For new samples, I have added a tutorial README.md file to guide users through what they need to know to implement code using this feature. For example, see [conditional_rendering](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/extensions/conditional_rendering)
- [ ] For new samples, I have added a link to the [Antora navigation](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/antora/modules/ROOT/nav.adoc) so that the sample will be listed at the Vulkan documentation site

gpx1000 committed:
Add a complex sample showing how to do a simplistic tool that shows a Voxel Point Cloud map using Octomap.  Also demonstrates how to use ImGUI in a project with added functionality and positioning of rendered assets.

SaschaWillems commented:
Wanted to give this a first try, but no matter what I tried, I wasn't able to get the newly added OctoMap submodule to checkout. Any ideas? Tried to clean reset, did a recursive checkout of all submodules, etc.

gary-sweet commented:
> Wanted to give this a first try, but no matter what I tried, I wasn't able to get the newly added OctoMap submodule to checkout. Any ideas? Tried to clean reset, did a recursive checkout of all submodules, etc.

Same problem for me it looks like.

gpx1000 committed:
Merge branch 'main' into complex-octmap-rendering

gpx1000 committed:
Add Gaussian splat rendering and GLTF support to octomap sample

Extends the octomap sample with three view modes: Octomap, GLTF Regular, and GLTF Splats. Implements splat rendering pipeline with per-instance attributes and premultiplied alpha blending. Adds GLTF mesh rendering with vertex color support. Includes new shader files for splat geometry and fragment evaluation. Updates camera to FirstPerson type for WASD movement. Fixes cube winding order and improves command buffer recording. Adds proper license headers to all new and modified files.

gpx1000 committed:
Reformat and reorganize includes in octomap sample

Standardizes code formatting with consistent indentation and brace style. Reorders includes alphabetically with core utilities first. Consolidates constructor initializer list to single line. Adjusts spacing and alignment throughout for consistency.

gpx1000 committed:
Add explicit submodule update step to all workflow jobs

Ensures submodules are fully initialized and updated in each job by adding explicit `git submodule update --init --recursive` command after checkout, complementing the existing `submodules: "recursive"` checkout option.

gpx1000 committed:
Fix indentation in .gitmodules for octomap submodule

Replaces spaces with tabs to maintain consistent formatting throughout the .gitmodules file.

gpx1000 committed:
Add octomap file for complex render sample

gpx1000 committed:
Restrict render_octomap sample to platforms with GLFW support

gpx1000 committed:
Force Windows to build library as static to hopefully fix CI.

gpx1000 committed:
Implement update_overlay method for render_octomap sample

Adds update_overlay method override that delegates to custom ImGUI implementation in render() function and executes additional UI callbacks when provided.

gpx1000 committed:
Fix header guard naming and exclude octomap submodule from CI checks

Renames header guards in ImGUIUtil.h from ORB_SLAM3_IMGUI_H to IMGUI_UTIL_H to match the actual file name. Excludes samples/complex/render_octomap/octomap directory from CI workflow file checks.

gpx1000 committed:
Add braces to single-statement conditionals and replace C-style casts with static_cast

Improves code style consistency by adding braces to all single-statement if blocks. Replaces C-style casts with appropriate static_cast and reinterpret_cast. Filters out third_party/glfw warnings from clang-tidy output in CI workflow.

gpx1000 committed:
Fix indentation in render_octomap sample

Replaces leading spaces with tabs to maintain consistent formatting throughout the file.

SaschaWillems commented:
This does compile for me now (after explicitly updating submodules), but crashes at startup with the following error:

"ERROR: Filestream to assets/scenes/octmap_and_splats/octMap.bin not open, nothing read."

I do see octMap.bin directly below the assets/scene folder.

gpx1000 commented:
Yeah, I was holding off on uploading the map until I took a good one that
includes the splats.  Will upload as soon as I can go to the office and
capture it

On Thu, Jan 1, 2026, 1:30 PM Sascha Willems ***@***.***>
wrote:

> *SaschaWillems* left a comment (KhronosGroup/Vulkan-Samples#1247)
> <https://github.com/KhronosGroup/Vulkan-Samples/pull/1247#issuecomment-3704128260>
>
> This does compile for me now (after explicitly updating submodules), but
> crashes at startup with the following error:
>
> "ERROR: Filestream to assets/scenes/octmap_and_splats/octMap.bin not open,
> nothing read."
>
> I do see octMap.bin directly below the assets/scene folder.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/KhronosGroup/Vulkan-Samples/pull/1247#issuecomment-3704128260>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA5IAY2XXH22KNPXLRU5TK34EWGY3AVCNFSM6AAAAABTTGIWPOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTOMBUGEZDQMRWGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

gary-sweet commented:
Not supported on direct-to-display platforms. Doesn't even build the sample.

gpx1000 committed:
Refactor ImGUI input handling to use platform-agnostic framework events

Replaces GLFW-specific key handling with framework's KeyCode enum for platform-agnostic input. Renames handleKey to handle_key_event to match framework naming conventions. Updates modifier key state tracking to use ImGui's key state. Removes GLFW dependency and platform-specific conditionals from ImGUI implementation. Enables render_octomap sample on all platforms including Android, iOS, and Direct-to-Display.

gpx1000 committed:
Update copyright year to 2026

Updates copyright headers across framework components and shaders from 2025 to 2026.

gpx1000 committed:
Update copyright years in project files

Updated copyright headers across various project files to reflect the year 2026. This ensures consistency and compliance with licensing requirements.

gpx1000 commented:
Give it another try Gary, I removed the input handling by GLFW assumption used during development.  Although, I *still* haven't had the time to generate a new map to go along with this; it should build for everyone on all devices and then run until you get to Sascha's problem of not having the map yet.

gary-sweet commented:
> Give it another try Gary, I removed the input handling by GLFW assumption used during development. Although, I _still_ haven't had the time to generate a new map to go along with this; it should build for everyone on all devices and then run until you get to Sascha's problem of not having the map yet.

Confirmed. This now gets to the same point for me.

gpx1000 committed:
Merge branch 'refs/heads/main' into complex-octmap-rendering

# Conflicts:
#	.gitmodules
#	framework/CMakeLists.txt

gpx1000 committed:
Update Gaussian splatting implementation to follow KHR_gaussian_splatting spec

Moves Gaussian splatting attributes from extension object to primitive attributes using KHR_gaussian_splatting namespace. Updates attribute parsing to check KHR_gaussian_splatting:ROTATION, SCALE, OPACITY, and SH_DEGREE_0_COEF_0 in primitive attributes first, with fallback to old draft extension properties and COLOR_0. Simplifies extension parsing logic and removes nested property structures. Updates code examples in tutorial documentation to reflect new attribute structure.

gpx1000 committed:
clang-format

gpx1000 committed:
fix the compile

gary-sweet commented:
Are these errors expected:

```
ERROR: Filestream to assets/scenes/octmap_and_splats/octMap.bin not open, nothing read.
[error] Error Message: Failed to open file for reading at path: shaders/render_octomap/glsl/imgui.vert.spv
[error] Failed when running application Octmap rendering
```

janharaldfredriksen-arm commented on samples/complex/render_octomap/README.adoc:
```diff
@@ -0,0 +1,46 @@
+
+The library https://octomap.github.io/[octomap] provides just such a library.
+
+In this sample, we demonstrate how using instancing in Vulkan, we can dynamically display and update an Octomap that was generated by ARKit.  Using MoltenVK, we can display the map in real time, and communicate it rapidly over a network for desktops to render and combine in real time.
```
Is this still using MoltenVK or is this out of date?

gpx1000 committed:
Merge branch 'main' into complex-octmap-rendering

gpx1000 committed:
Update to latest KHR and use sample map.

gpx1000 committed:
Update to show we do not need to use MoltenVK for rendering.

gpx1000 committed:
clang-format fixes.

gpx1000 commented:
@KhronosGroup/vulkan-samples-reviewers Finally ready for review.

gpx1000 committed:
Fix windows having 32bit limit on a static lib.

asuessenbach commented on samples/complex/render_octomap/render_octomap.cpp:
```diff
@@ -913,10 +937,70 @@ void render_octomap::loadGLTFScene(const std::string &filename)
+		if (!entry.is_regular_file())
+			continue;
+		std::string name = entry.path().filename().string();
+		if (name.find("_cell_") != std::string::npos && name.size() >= 5 &&
```
If `name` contains "`_cell_`", its size is `>= 5`. So you don't need to check that.

asuessenbach commented on samples/complex/render_octomap/render_octomap.cpp:
```diff
@@ -913,10 +937,70 @@ void render_octomap::loadGLTFScene(const std::string &filename)
+			continue;
+		std::string name = entry.path().filename().string();
+		if (name.find("_cell_") != std::string::npos && name.size() >= 5 &&
+		    name.substr(name.size() - 5) == ".gltf")
```
Maybe `name.ends_with( ".gltf" )`?

gpx1000 replied:
ends_with is C++20.  I don't know if we can use it.  I think we have a max upper bound on C++ that we can use and I don't remember off the top of my head if it was above 20 or not.

SaschaWillems replied:
We upped that to C++20 several months ago for all platforms, so should be good.

gpx1000 replied:
I thought we had to downgrade it again as there was a max version of gcc that we were allowed to use as some couldn't upgrade their developer systems.

gary-sweet commented:
Here's what I get when I run this:

```
Reading binary octree type OcTree
[error] Error Message: Failed to open file for reading at path: shaders/render_octomap/glsl/imgui.vert.spv
[error] Failed when running application Octmap rendering
[error] -1363947098 - VUID-vkDestroyPipeline-pipeline-parameter: vkDestroyPipeline(): pipeline Invalid VkPipeline Object 0x3f401698bf866666.
The Vulkan spec states: If pipeline is not VK_NULL_HANDLE, pipeline must be a valid VkPipeline handle (https://docs.vulkan.org/spec/latest/chapters/pipelines.html#VUID-vkDestroyPipeline-pipeline-parameter)
vulkan_samples: /build/gsweet/vc5_git/thirdparty_vulkan/Vulkan-ValidationLayers/layers/./thread_tracker/thread_safety_validation.h:122: std::shared_ptr<threadsafety::ObjectUseData> threadsafety::Counter<T>::FindObject(T, const Location&) [with T = VkPipeline_T*]: Assertion `object_table.contains(object)' failed.
```
It looks like there are no .spv files in the shaders/render_octomap/glsl folder at all currently.

gary-sweet commented:
If I build the spv files manually it does run, but I'm not sure it looks right.

My normal method of driving the UI (using the keyboard) doesn't work either so I can't actually do anything other than show the initial result.

This is what I see:

<img width="1280" height="720" alt="render_octomap-1970-01-01---00-30-33" src="https://github.com/user-attachments/assets/33fdf146-bb59-406b-9002-d6a83bed7f76" />

asuessenbach reviewed (commented):
I don't get any octomap asset files.
When I checkout this PR, switch to branch 'complex-octomap-rendering', and update the asset submodule, everything seems to run fine, but there is no `octmap_and_splats` in `assets/scenes`. Any idea how to get that?

gpx1000 commented:
> If I build the spv files manually it does run, but I'm not sure it looks right.
> 
> My normal method of driving the UI (using the keyboard) doesn't work either so I can't actually do anything other than show the initial result.
> 
> This is what I see:
> 
> <img alt="render_octomap-1970-01-01---00-30-33" width="1280" height="720" src="https://private-user-images.githubusercontent.com/20819159/605655329-33fdf146-bb59-406b-9002-d6a83bed7f76.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODEwODM0MTcsIm5iZiI6MTc4MTA4MzExNywicGF0aCI6Ii8yMDgxOTE1OS82MDU2NTUzMjktMzNmZGYxNDYtYmI1OS00MDZiLTkwMDItZDZhODNiZWQ3Zjc2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA2MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNjEwVDA5MTgzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU3NWU3OWI4Njc5Y2JlZGM0ODVmMTMwZTdkNzI5MmFkNTllN2I3NDA5ZmMyODUwZThjNTMzYTVhYjg0YzIyMmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.Xf5cSD5gnIiUtiCJhxh3-7LbKb8MawEFqgSaQLV12LU">

Yep, that worked.  WASD and the mouse will let you move around.  That's the voxel view of the occupancy grid.  Click on the other two buttons to see other views of the same map.

gpx1000 commented:
> I don't get any octomap asset files. When I checkout this PR, switch to branch 'complex-octomap-rendering', and update the asset submodule, everything seems to run fine, but there is `no octmap_and_splats` in `assets/scenes`. Any idea how to get that?

It should just be in assets/scenes/octmap_and_splats  There should be 21 files in there.  Here's the  assets commit if you want to grab it directly: https://github.com/KhronosGroup/Vulkan-Samples-Assets/compare/8db8ce9c528330f0b1261b07531b009732b08731...6900e07d1ba4e431d311bb387edd3623061638bf

gary-sweet commented:
> Yep, that worked. WASD and the mouse will let you move around. That's the voxel view of the occupancy grid. Click on the other two buttons to see other views of the same map.

But I can't click on the buttons (we don't have a mouse). Normally the arrow keys move around in the UI, but not in this sample it seems.

gpx1000 commented:
> But I can't click on the buttons (we don't have a mouse). Normally the arrow keys move around in the UI, but not in this sample it seems.

Okay, gimme a bit and there'll be a new version that lets you move around arrow keys as normal.  The complex sample I original designed to use less of the framework so it doesn't inherit the input.  Easy fix working on it and the shader compiler now.

gpx1000 committed:
Allow imgui to update with keyboard nav.  fix name.size() >= 5 and add in built spv files so platforms that oddly don't auto build will work.

gpx1000 commented:
> > Yep, that worked. WASD and the mouse will let you move around. That's the voxel view of the occupancy grid. Click on the other two buttons to see other views of the same map.
> 
> But I can't click on the buttons (we don't have a mouse). Normally the arrow keys move around in the UI, but not in this sample it seems.

This version should allow you to do keyboard nav in imgui.

gary-sweet commented:
> This version should allow you to do keyboard nav in imgui.

Sadly not:

```
Reading binary octree type OcTree
    01:46:01.620 nexus_display_hlm: usage 6 enabled
vulkan_samples: /build/gsweet/vc5_git/thirdparty_vulkan/Vulkan-Samples/third_party/imgui/imgui.cpp:1488: void ImGuiIO::AddKeyAnalogEvent(ImGuiKey, bool, float): Assertion `KeyMap[n] == -1 && "Backend needs to either only use io.AddKeyEvent(), either only fill legacy io.KeysDown[] + io.KeyMap[]. Not both!"' failed.
Aborted
```

SaschaWillems commented:
Sample does run for me, but I'm seeing a similar issue to the one Gary reported. As soon as I press a key (e.g. "S") I get an assertion:

<img width="412" height="328" alt="image" src="https://github.com/user-attachments/assets/4d15dac7-b95a-48bf-b001-53f7eb3409cc" />

Also unsure what it's supposed to look like. Can you post/add some reference screenshots?

gpx1000 committed:
Fix for keyboard navigation and numpad look handling

gpx1000 commented:
<img width="1275" height="777" alt="image" src="https://github.com/user-attachments/assets/8e0ad64d-8f23-4586-ad3a-8cf3e4b44683" />
<img width="1275" height="777" alt="image" src="https://github.com/user-attachments/assets/cdfe58f6-0f39-4090-8cac-f20bf1b40ec7" />
<img width="1275" height="777" alt="image" src="https://github.com/user-attachments/assets/32f98af3-f3dc-4b4b-8c96-8ad678f1cfe2" />

This is a temp map while I make a new one that hopefully looks nicer.  The gaussian splats are from a single live run so they don't look that detailed.  They are accurate.  The octomap renders are as described in the text, no textures, just points in an occupancy grid.  So those are by design false colors.

SaschaWillems commented on samples/complex/render_octomap/render_octomap.cpp:
```diff
@@ -0,0 +1,1224 @@
+#include <tiny_gltf.h>
+
+// KHR_gaussian_splatting extension name
+#define KHR_GAUSSIAN_SPLATTING_EXTENSION "KHR_gaussian_splatting"
```
Minor nitpick, but this seems to be a glTF extension (not a Vulkan one), so maybe prefix the define with GLTF to avoid confusion?

gpx1000 committed:
clang-format

SaschaWillems commented:
Could documentation be expanded at least a bit? The code has very few comments and the readme is mostly talking about the topic and not the sample itself.

gpx1000 committed:
address Sascha's nit on KHR_gaussian_splatting extension name  (Diverges from what we do in GLTF_Loader in the framework, but it's a good idea).

gary-sweet commented:
The UI traversal is now working, thanks.

gpx1000 committed:
Expand README for Octomap sample to include usage, controls, rendering modes, and key implementation details.

gpx1000 commented:
> Could documentation be expanded at least a bit? The code has very few comments and the readme is mostly talking about the topic and not the sample itself.

I updated the README.adoc.  If you'd like specific information let me know and I'll add it.  I think I'm headed to bed for a nap so might be a few hours before I do another update.

SaschaWillems commented:
That's perfect. Thank you very much 👍🏻

r-potter commented:
No reflection on the quality of the sample but we should hold off merging this until internal discussions are fully resolved (but please continue to review).

gpx1000 committed:
I forgot to add in the complex samples section to nav.adoc along with the rest of the documentation.

asuessenbach commented on samples/complex/render_octomap/README.adoc:
```diff
@@ -17,22 +17,132 @@
-The preceding high-level background describes how we get to a situation where there's plenty of desire to be able to work with and render point clouds that are dynamically generated.  ARCore, and ARKit are both able to create a point cloud map, and everything from Drones to robots use this same basic system to register and deal with the world around them.  To navigate a room, a drone/robot might  ...
+The preceding high-level background describes how we get to a situation where there's plenty of desire to be able to work with and render point clouds that are dynamically generated. ARCore and ARKit are both able to create a point cloud map, and everything from drones to robots use this same basic system to register and deal with the world around them. To navigate a room, a drone/robot might nee ...
 
 The library https://octomap.github.io/[octomap] provides just such a library.
```
"The library ... provides just such a library"?

Maybe better: "... provides such functionality"?

gpx1000 committed:
address comment.

marty-johnson59 commented:
Hi @asuessenbach, are you OK if we merge this PR?  It has fixes to CI that are breaking things, so would like to do this soon (or else spin the fixes out separately..).  LMK, thanks!

asuessenbach commented:
> Hi @asuessenbach, are you OK if we merge this PR? It has fixes to CI that are breaking things, so would like to do this soon (or else spin the fixes out separately..). LMK, thanks!

@marty-johnson59 I hope, I can check this PR this week again.
But anyways, I'd prefer to have fixes to CI in a separate PR. Even if it's just to not hide them in some lengthy, unrelated new sample.

asuessenbach commented on .github/workflows/check.yml:
```diff
@@ -29,7 +29,7 @@ jobs:
           git config --global --add safe.directory /__w/Vulkan-Samples/Vulkan-Samples
           git fetch origin $TARGET_BRANCH:$TARGET_BRANCH
-          echo all="$(git diff --name-only --diff-filter=ACMRT $TARGET_BRANCH | grep -v -e .github -e third_party -e .in$ | xargs)" >> $GITHUB_OUTPUT
+          echo all="$(git diff --name-only --diff-filter=ACMRT $TARGET_BRANCH | grep -v -e .github -e third_party -e 'samples/complex/render_octomap/octomap' -e .in$ | xargs)" >> $GITHUB_OUTPUT
```
Why is that octomap stuff located in the samples folder, not in the third_party folder?

gpx1000 replied:
I didn't think any sample outside of this octomap sample would want to use it.  Thus, I contained it similar to how we did the OpenCL sample.

asuessenbach replied:
And you're absolutely sure that no sample will ever use that stuff?

But I don't think that really matters. I'd just prefer it if third-party files were stored in the “third_party” folder.
And I wasn't aware of any examples where third-party files are already stored locally. Maybe we should move those files accordingly?

gpx1000 replied:
I moved it as requested.

asuessenbach commented on .github/workflows/build.yml:
Should the changes here be moved into a separate PR? They are not directly related to this sample.

gpx1000 replied:
These are specific to this sample since we have submodules in this folder as well.

asuessenbach commented on bldsys/cmake/sample_helper.cmake:
```diff
@@ -155,7 +155,7 @@ function(add_project)
 
 if(${TARGET_TYPE} STREQUAL "Sample")
-    add_library(${PROJECT_NAME} OBJECT ${TARGET_FILES} ${SHADERS_GLSL} ${SHADERS_HLSL} ${SHADERS_SLANG} ${SHADERS_SPVASM})
+    add_library(${PROJECT_NAME} STATIC ${TARGET_FILES} ${SHADERS_GLSL} ${SHADERS_HLSL} ${SHADERS_SLANG} ${SHADERS_SPVASM})
```
Should this be moved into a separate PR?

gpx1000 replied:
#1550 now has this fix.  It is what fixes the CI windows building issue.

asuessenbach commented on framework/scene_graph/components/gaussian_splat.h:
```diff
@@ -0,0 +1,101 @@
+ * - Opacity (alpha value)
+ * - Color (RGB or spherical harmonics coefficients)
+ */
+class GaussianSplat : public Component
```
It seems, the `GaussianSplat` class is never used?

asuessenbach commented on samples/complex/render_octomap/CMakeLists.txt:
```diff
@@ -0,0 +1,75 @@
+        NAME "Octmap rendering"
+        DESCRIPTION "Demonstration of how to render an OctMap which was generated from ICP combining ARCore and ARKit SLAM maps"
+        FILES
+            ImGUIUtil.cpp
```
Maybe add `ImGUIUtil.h` here as well.

asuessenbach commented on samples/complex/render_octomap/Screens/MapView.cpp:
```diff
@@ -0,0 +1,182 @@
+std::vector<VkWriteDescriptorSet> MapView::LoadAssets(ApiVulkanSample *, const VkDescriptorSetAllocateInfo &, VkQueue)
+{
+	return {};
+}
```
Is this function supposed to be filled with content, at some point?

gpx1000 replied:
It's meant to provide a template for end users as this is where they can load jpg png or other assets for buttons in the GUI.

asuessenbach replied:
Well, in that case I would add at least some dummy asset to showcase how a jpg or png maps to one or more `VkWriteDescriptorSet`.
Or just remove this function and add some comment in `ImGUIUtil::initResources` where assets for buttons and such should be loaded.

asuessenbach commented on samples/complex/render_octomap/Screens/MapView.h:
```diff
@@ -0,0 +1,54 @@
+	bool DrawUI();
+
+  private:
+	void DrawSidebar();
```
Neither `DrawUI`, nore `DrawSidebar` are ever called.

asuessenbach commented on samples/complex/render_octomap/Tutorials/capturing-gaussian-splats.adoc:
```diff
@@ -0,0 +1,717 @@
+
+== Capture Method 2: SLAM (Simultaneous Localization and Mapping)
+
+SLAM is what robots and AR devices use to understand their environment in real-time. Unlike photogrammetry (which processes everything after capture), SLAM tracks the camera pose *while* you're capturing. This is what the `render_octomap` sample generated by Holochip's real-time SLAM.  Note, Holochip's solution generates the GLTF with compressed, embedded Gaussian Splats in real-time without any  ...
```
> This is what the `render_octomap` sample generated by Holochip's real-time SLAM.

Missing verb?

asuessenbach commented on samples/complex/render_octomap/Tutorials/capturing-gaussian-splats.adoc:
```diff
@@ -0,0 +1,717 @@
+**Hardware:**
+
+- A camera with good frame rate (30+ FPS)
+- Ideally, an RGB-D camera (color + depth) like:
```
like what?

asuessenbach reviewed (commented):
Switching between the three modes seems to also change the position and/or view direction. Is that right?

gpx1000 committed:
Address feedback.

gpx1000 commented:
> Switching between the three modes seems to also change the position and/or view direction. Is that right?

Yep, this is expected as the maps are not geolocated.  I didn't include the GEOJSON code that handles the geolocation figuring that was beyond the scope.

asuessenbach commented on samples/complex/render_octomap/Tutorials/gaussian-splats-rendering.adoc:
```diff
@@ -0,0 +1,1290 @@
+[stem]
+++++
+G(x) = e^{-\frac{1}{2}(x-\mu)^T \Sigma^{-1} (x-\mu)}
+++++
```
I assume, this formular is supposed to be formatted, somehow?
I just see it as
` \$G(x) = e^{-\frac{1}{2}(x-\mu)^T \Sigma^{-1} (x-\mu)}\$`.
Is there anything, I could do to see it formatted?

gpx1000 replied:
LaTeX version is now updated.  lemme know if you need an Adoc renderer that handles LaTeX plugins.

asuessenbach replied:
Yes, please, what adoc renderer would you recommend?

asuessenbach commented on samples/complex/render_octomap/Tutorials/gaussian-splats-rendering.adoc:
```diff
@@ -0,0 +1,1290 @@
+* `antialiased`: Enable anti-aliasing filter
+* `colorSpace`: Color space ("BT.709-sRGB" or "linear")
+* `kernel`: Splat kernel type ("ellipse" or "sphere")
+* `shDegree`: Spherical harmonics degree (0-3)
```
`antialiased` and `shDegree` are not listed in the extension structure above.

asuessenbach commented on samples/complex/render_octomap/Tutorials/gaussian-splats-rendering.adoc:
```diff
@@ -0,0 +1,1290 @@
+Here's the real story about compression in the GLTF standard: the `KHR_gaussian_splatting` extension works alongside a companion extension called `KHR_gaussian_splatting_compression_spz` that handles compression using the SPZ format.
+
+SPZ (Splat compression format) was specifically designed for Gaussian splats and achieves up to **90% compression compared to uncompressed PLY** while preserving visual fidelity and performance. This isn't a repurposed mesh compression format - it's purpose-built for splat data.
+
```
What's that `PLY` mentioned here as uncompressed?

gpx1000 replied:
Fixed in the text.  However, PLY is the uncompressed version.  There's many different compression ideas and the leading one is PLZ which we demonstrate here and which the KHR in glTF also supports (although strictly speaking the KHR supports any type of compression format).

asuessenbach commented on samples/complex/render_octomap/Tutorials/gaussian-splats-rendering.adoc:
```diff
@@ -0,0 +1,1290 @@
+        "KHR_gaussian_splatting:SH_DEGREE_0_COEF_0": 1,
+        "KHR_gaussian_splatting:ROTATION": 2,
+        "KHR_gaussian_splatting:SCALE": 3,
+        "KHR_gaussian_splatting:OPACITY": 4
```
Why is the order of attributes different than in the GLTF structure above, with just `KHR_gaussian_splatting`.

asuessenbach commented on samples/complex/render_octomap/Tutorials/gaussian-splats-rendering.adoc:
```diff
@@ -0,0 +1,1290 @@
+
+[source,cpp]
+----
+class GaussianSplat : public Component
```
The GaussianSplat component is gone.

asuessenbach commented on samples/complex/render_octomap/Tutorials/gaussian-splats-rendering.adoc:
```diff
@@ -0,0 +1,1290 @@
+
+    return true;
+}
+----
```
None of all those code snippets are found in the actual sample, right?
Why are you doing it differently in the sample?

gpx1000 replied:
These should now better reflect the sample.

asuessenbach commented on samples/complex/render_octomap/Tutorials/gaussian-splats-rendering.adoc:
```diff
@@ -0,0 +1,1290 @@
+
+**Step 3 - Fragment Shader:** For each pixel, we evaluate the 2D Gaussian function. This gives us a smooth falloff from the center to the edges. We multiply by the opacity and color, output premultiplied alpha, and we're done.
+
+**Step 4 - Blending:** The GPU blends all the quads together using alpha blending. Since we're outputting premultiplied alpha, the standard blend equation `src + dst * (1 - src.alpha)` does exactly what we want.
```
I'm not sure what you mean with "standard blend equation"... but I think, it should be `src * src.alpha + dst * (1 - src.alpha)`?

And, the "standard blend problem", this looks correct only if the fragments are sorted from far to near.

gpx1000 replied:
It's premultiplied; I fixed it in the most recent version.

asuessenbach commented on samples/complex/render_octomap/Tutorials/gaussian-splats-rendering.adoc:
```diff
@@ -0,0 +1,1290 @@
+
+=== The Projection Math (Don't Worry, I'll Explain)
+
+Here's the tricky part: we have a 3D Gaussian blob floating in space with its 3D covariance matrix `Σ₃D`. We need to project it onto the 2D screen and figure out what its 2D covariance matrix `Σ₂D` should be. If we get this wrong, the splat will look stretched or squashed on screen.
```
Can the `Σ₃D` be formatted to have the 'D' as a subscript, just like the '3'?

gpx1000 replied:
I switched to LaTeX everywhere.

asuessenbach commented on samples/complex/render_octomap/Tutorials/gaussian-splats-rendering.adoc:
```diff
@@ -0,0 +1,1290 @@
+
+[source,cpp]
+----
+template<typename T>
```
The template parameter `T` has to be `float`, or some multi-`float `type!

gpx1000 replied:
Added a comment to enforce the float.   I'm not sure if there's anything else we can do to make that enforcement more robust.

asuessenbach replied:
Maybe just return a `std::vector<float>`, filled with `accessor.count * componentCount( accessor.type )` `float`s, where `componentCount()` is some translation from the `type` string to the actual number of components.

gpx1000 committed:
Update Gaussian Splat documentation and implementation to include additional attributes and clarifications

- Add `:stem:` attribute for mathematical expressions in the documentation.
- Clarify SPZ compression details and the use of PLY as raw data storage.
- Adjust JSON examples to include new attributes (`antialiased` and `shDegree`).
- Switch to `SplatInstance` struct for interleaved data storage in the code.
- Expand explanations around projection math, blending, and premultiplied alpha use.
- Highlight mandatory depth sorting for correct rendering.

gpx1000 committed:
Update to better describe how to add custom UI textures.

gpx1000 committed:
clang-format

asuessenbach commented on samples/complex/render_octomap/ImGUIUtil.h:
```diff
@@ -0,0 +1,105 @@
+	enum ViewState
+	{
+		LIVEMAPS_ACTIVE,
+	} state;
```
None of `montserratExtraBoldNormal`, `montserratExtraBoldSmall`, `montserratBoldNormal`, `montserratRegularNormal`, `selectedStyle`, `windowWidth`, `windowHeight`, or the `enum ViewState` is ever used.

asuessenbach commented on samples/complex/render_octomap/ImGUIUtil.h:
```diff
@@ -0,0 +1,105 @@
+	ImFont                               *montserratRegularNormal;
+	int                                   selectedStyle = 0;
+	float                                 windowWidth, windowHeight;
+	bool                                  needsUpdateBuffers = false;
```
You're mixing `camelCase` and `snake_case` for the member variables. I think, `snake_case` is the standard here.

asuessenbach commented on samples/complex/render_octomap/ImGUIUtil.h:
```diff
@@ -0,0 +1,105 @@
+		LIVEMAPS_ACTIVE,
+	} state;
+
+	MapView MapsView;
```
Having a member variable start with an uppercase char is somewhat unexpected. Maybe rename it to `map_view`, or so?

asuessenbach commented on samples/complex/render_octomap/ImGUIUtil.h:
```diff
@@ -0,0 +1,105 @@
+
+	static bool GetWantKeyCapture();
+
+	static void charPressed(uint32_t key);
```
Member functions are also `snake_case` in this repo.

asuessenbach commented on samples/complex/render_octomap/README.adoc:
```diff
@@ -0,0 +1,156 @@
+* **Push constants for scale/translate** — the pipeline uses a small push constant block (`PushConstBlock`) containing a 2D scale and translate, allowing the UI to be mapped to any sub-region of the framebuffer without a separate uniform buffer.
+* **Image rendering in the GUI** — to display a Vulkan render target inside ImGui, a `VkDescriptorSet` is allocated for the target's `VkImageView`, then passed to `ImGui::Image()` as an `ImTextureID`. This is the standard pattern for embedding Vulkan render targets in ImGui and is demonstrated here as a reusable utility.
+* **Font management** — the wrapper loads custom Montserrat fonts (ExtraBold, Bold, Regular) at multiple sizes and uploads them as a single font atlas texture, demonstrating how to use custom fonts with a Vulkan ImGui backend.
+
```
As noted above, none of the `ImFonts` (`montserrat*`) is used here.

asuessenbach commented on samples/complex/render_octomap/README.adoc:
```diff
@@ -0,0 +1,156 @@
+
+Compare the file sizes of the `.bt` octomap and the splat GLTF to see the storage trade-off: the octomap is orders of magnitude smaller because it stores only occupied-voxel positions and sizes, whereas a GLTF encodes full geometry, normals, UVs, and textures for every surface. This difference is critical for drones and ROS-powered devices that may not have enough storage for GLTF-scale maps of a ...
+
+==== 3. Custom ImGui Integration with Descriptor Sets for Images (`ImGUIUtil.cpp` / `ImGUIUtil.h`)
```
I don't think the ImGui integration is an integral part of the Octomap rendering example. It's simply the first example that requires or uses this additional functionality.
Would it be worth to move that integration over to the framework?

gpx1000 replied:
The original idea was this was to demonstrate a multiple system setup.  The reasoning for ImGUI being pulled into the sample instead of the framework is to provide a direct "this is how to use it" that doesn't bury that implementation.  It's to make it trivial for users to pull an example ImGui into their own projects.  I would prefer to leave it here for now and see if we get feedback that validates if the complex sample idea is a valid path forward or not (i.e. if users think that demonstrating multiple systems working together without a framework that might be harder to read is of benefit).

asuessenbach replied:
Ok... But why did you select the ImGui out of the framework? You're using other parts of it as well.

And I'd call additional stuff in a sample that might hide the actually important code distracting. If the ImGui handling is somehow important for the octomap rendering I'd be fine with it. But I don't see that.

gpx1000 replied:
Well the ImGui in the framework doesn't have how to use it with custom textures.  However, I guess that isn't much of a feature to showcase.

If it's distracting from the octomap / gaussian splat instruction / rendering, then I'll remove it and use the framework's ImGui.

gpx1000 replied:
I removed ImGui and moved back to the Framework's ImGui.

asuessenbach commented on samples/complex/render_octomap/render_octomap.cpp:
```diff
@@ -0,0 +1,1244 @@
+
+render_octomap::render_octomap() :
+    vertices(), indexCount(), pipelineCache(VK_NULL_HANDLE), pipelineLayout(VK_NULL_HANDLE), pipeline(VK_NULL_HANDLE), descriptorPool(VK_NULL_HANDLE), descriptorSetLayout(VK_NULL_HANDLE), descriptorSet(VK_NULL_HANDLE), gui(nullptr), mMaxTreeDepth(16), m_zMin(), m_zMax(), lastMapBuildSize()
+{
```
I don't like lines with 291 columns.

Besides that, simple initializations like that should be done via default member initializers.

asuessenbach commented on samples/complex/render_octomap/render_octomap.cpp:
```diff
@@ -0,0 +1,1244 @@
+	{
+		return;
+	}
+	float nextBuildSize = static_cast<float>(lastMapBuildSize) + (static_cast<float>(lastMapBuildSize) * 0.05f);
```
`float nextBuildSize = lastMapBuildSize * 1.05f`?

And: what does this check mean?

asuessenbach commented on samples/complex/render_octomap/render_octomap.cpp:
```diff
@@ -0,0 +1,1244 @@
+	m_zMax = static_cast<float>(maxZ);
+
+	// this is to get just grey; doing full color heightmap for now.
+	// h = std::min(std::max((h-m_zMin)/ (m_zMax - m_zMin), 0.0f), 1.0f) * 0.4f + 0.3f; // h \in [0.3, 0.7]
```
Is it important to keep this out-commented code?

asuessenbach commented on samples/complex/render_octomap/render_octomap.cpp:
```diff
@@ -0,0 +1,1244 @@
+			}
+			else
+			{
+				h = (1.0f - std::min(std::max((h - m_zMin) / (m_zMax - m_zMin), 0.0f), 1.0f)) * 0.8f;
```
Maybe use `std::clamp()` for such `std::min(std::max())` situations.

And I don't know how performance sensitive this is (or how aggressive compilers optimize), but some `float zRange = m_zMax - m_zMin`, calculated once out of the loop, could help here.

Moreover: some magic `0.8f`? Where does it come from and what does it mean?

asuessenbach commented on samples/complex/render_octomap/render_octomap.cpp:
```diff
@@ -0,0 +1,1244 @@
+					g = 0.5f;
+					b = 0.5f;
+					break;
+			}
```
Is this the standard hsv to rgb conversion?
Then you could introduce a little helper function for it.

asuessenbach commented on samples/complex/render_octomap/render_octomap.cpp:
```diff
@@ -0,0 +1,1244 @@
+		vkCmdBindIndexBuffer(draw_cmd_buffers[i], indexBuffer->get_handle(), 0, VK_INDEX_TYPE_UINT32);
+
+		vkCmdDrawIndexed(draw_cmd_buffers[i], indexCount, instances.size(), 0, 0, 0);
+		// draw_ui(draw_cmd_buffers[i]);
```
Need to keep this out-commented code?

asuessenbach commented on samples/complex/render_octomap/render_octomap.cpp:
```diff
@@ -0,0 +1,1244 @@
+	return true;
+}
+
+void render_octomap::createPipelines(VkRenderPass renderPass)
```
`create_gltf_pipeline` creates the gltf pipeline; `create_splat_pipeline` creates the splat pipeline. But what pipelines does `createPipelines` create?
Besides that: it's camelCase instead of snake_case.

asuessenbach commented on samples/complex/render_octomap/render_octomap.cpp:
```diff
@@ -0,0 +1,1244 @@
+		auto buf = instanceBuffer->map();
+		memcpy(buf, instances.data(), instances.size() * sizeof(InstanceData));
+		instanceBuffer->flush();
+		instanceBuffer->unmap();
```
Should the instance buffer (as well as the vertex buffer and the index buffer) be located on the GPU? That is, shouldn't we use a staging buffer to copy the data over to the GPU?

asuessenbach commented on samples/complex/render_octomap/render_octomap.cpp:
```diff
@@ -0,0 +1,1244 @@
+			vkCmdBindDescriptorSets(cmd, VK_PIPELINE_BIND_POINT_GRAPHICS, pipelineLayout, 0, 1, &descriptorSet, 0, nullptr);
+			vkCmdBindPipeline(cmd, VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline);
+			vkCmdBindVertexBuffers(cmd, 0, 1, &vertexBuffer->get_handle(), offsets);
+			if (instanceBuffer)
```
Why do you check for `instanceBuffer`, but not for `indexBuffer`?

asuessenbach commented on samples/complex/render_octomap/render_octomap.cpp:
```diff
@@ -0,0 +1,1244 @@
+			{
+				create_splat_pipeline(render_pass);
+			}
+			if (splat_pipeline == VK_NULL_HANDLE)
```
Does this mean, failing to create the `splat_pipeline` is an acceptable behaviour?

gpx1000 replied:
We should assert now.

gpx1000 committed:
Address Andreas' comments.  Thanks!  keep them coming!

gpx1000 committed:
clang-format.

gpx1000 committed:
try to resolve clang-format

asuessenbach commented on samples/complex/render_octomap/render_octomap.h:
```diff
@@ -95,20 +95,20 @@ class render_octomap : public ApiVulkanSample
+	unsigned int                        mMaxTreeDepth{16};
+
+	float                                              m_zMin{0.0f};
+	float                                              m_zMax{0.0f};
```
None of the member variables are prefixed with `m_`. But these two are.

asuessenbach commented on samples/complex/render_octomap/ImGUIUtil.cpp:
```diff
@@ -0,0 +1,841 @@
+{
+	// This conflicts with the hpp_gui context.  Disable for now.
+	// ImGuiContext* context = ImGui::CreateContext();
+	// ImGui::SetCurrentContext(context);
```
What conflict do you have with the `hpp_gui context` here?
And.. what's the `hpp_gui context` at all?

asuessenbach commented on samples/complex/render_octomap/ImGUIUtil.cpp:
```diff
@@ -0,0 +1,841 @@
+	        .with_usage(VK_BUFFER_USAGE_INDEX_BUFFER_BIT)
+	        .with_vma_usage(VMA_MEMORY_USAGE_GPU_TO_CPU)
+	        .with_debug_name("GUI index buffer")
+	        .build_unique(device);
```
A vertex buffer and an index buffer with the size of 1?
Will be reset at the first updateBuffers anyways, so you could leave them default initialized and skip the initialization here.

asuessenbach commented on samples/complex/render_octomap/ImGUIUtil.cpp:
```diff
@@ -0,0 +1,841 @@
+
+ImGUIUtil::~ImGUIUtil()
+{
+	// ImGui::DestroyContext(); // this would double free due to the default one in hpp_gui...
```
`hpp_gui`, again? What does it mean?

asuessenbach commented on samples/complex/render_octomap/ImGUIUtil.cpp:
```diff
@@ -0,0 +1,841 @@
+}
+
+// Initialize styles, keys, etc.
+void ImGUIUtil::init(float width, float height)
```
Is there any reason why `init` and `initResources` are not part of the constructor?

asuessenbach commented on samples/complex/render_octomap/ImGUIUtil.cpp:
```diff
@@ -0,0 +1,841 @@
+	io.ConfigFlags |= ImGuiConfigFlags_NavEnableKeyboard;
+}
+
+void ImGUIUtil::setStyle(uint32_t index)
```
Is `setStyle` ever called, besides the one call from `init`? Is it needed, then?

asuessenbach commented on samples/complex/render_octomap/ImGUIUtil.cpp:
```diff
@@ -0,0 +1,841 @@
+	VkDeviceSize uploadSize = texWidth * texHeight * 4 * sizeof(char);
+
+	// SRS - Get Vulkan device driver information if available, use later for display
+	if (base->get_render_context().get_device().get_gpu().is_extension_supported(VK_KHR_DRIVER_PROPERTIES_EXTENSION_NAME))
```
Did you want to ask for
`	if (base->get_render_context().get_device().is_extension_enabled(VK_KHR_DRIVER_PROPERTIES_EXTENSION_NAME))` ?
If so, you should use `VkPhysicalDeviceDriverPropertiesKHR` below.

If you want to use `VkPhysicalDeviceProperties2`, you could check for `VK_VERSION_1_1` or higher. But that's the default with Vulkan-Samples.

asuessenbach commented on samples/complex/render_octomap/ImGUIUtil.cpp:
```diff
@@ -0,0 +1,841 @@
+	{
+		VkPhysicalDeviceProperties2 deviceProperties2 = {};
+		deviceProperties2.sType                       = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROPERTIES_2;
+		deviceProperties2.pNext                       = &driverProperties;
```
You're asking for something here that's never used.

asuessenbach commented on samples/complex/render_octomap/ImGUIUtil.cpp:
```diff
@@ -0,0 +1,841 @@
+
+		auto command_buffer = base->get_render_context().get_device().get_command_pool().request_command_buffer();
+
+		vkb::FencePool fence_pool{base->get_render_context().get_device()};
```
The `fence_pool` is never used.

asuessenbach commented on samples/complex/render_octomap/ImGUIUtil.cpp:
```diff
@@ -0,0 +1,841 @@
+		// Wait for the command buffer to finish its work before destroying the staging buffer
+		VK_CHECK(base->get_render_context().get_device().get_fence_pool().wait());
+		base->get_render_context().get_device().get_fence_pool().reset();
+		base->get_render_context().get_device().get_command_pool().reset_pool();
```
Resetting the fence pool and the command pool might be a bit too much. You don't know if some other fences or command buffers are still in flight.

asuessenbach commented on samples/complex/render_octomap/ImGUIUtil.cpp:
```diff
@@ -0,0 +1,841 @@
+	// Descriptor pool
+	std::vector poolSizes = {
+	    vkb::initializers::descriptor_pool_size(VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER, 6)};
+	VkDescriptorPoolCreateInfo descriptorPoolInfo = vkb::initializers::descriptor_pool_create_info(poolSizes, 7);
```
Magic numbers 6 and 7? Where do they come from, what do they mean?

asuessenbach commented on samples/complex/render_octomap/ImGUIUtil.cpp:
```diff
@@ -0,0 +1,841 @@
+	}
+}
+
+void ImGUIUtil::text_color_align(int align, const ImVec4 &col, const char *text, ...)
```
This function is never called.

asuessenbach commented:
There are still lots of variables and functions using camelCase, instead of snake_case.

Should we change them all, or do we want to start to ignore casing?

gpx1000 commented:
I'll go ahead and switch them to all camelCase.  Thanks for the review.  Working on fixes now.

gpx1000 committed:
Remove ImGui from the complex render_octomap sample inlcuding the instruction.

Move to the Framework imgui

Address rest of Andreas' comments.

asuessenbach commented on samples/complex/render_octomap/render_octomap.h:
```diff
@@ -0,0 +1,175 @@
+
+}        // namespace vkb
+
+class render_octomap : public ApiVulkanSample
```
Sorry for being so picky, but I think for classes we're using PascalCase:
render_ocotmap -> RenderOctomap

asuessenbach commented on samples/complex/render_octomap/render_octomap.cpp:
```diff
@@ -0,0 +1,1167 @@
+	std::vector dynamicStateEnables = {
+	    VK_DYNAMIC_STATE_VIEWPORT,
+	    VK_DYNAMIC_STATE_SCISSOR,
+	    VK_DYNAMIC_STATE_LINE_WIDTH};
```
Is the line width set anywhere?

asuessenbach commented on samples/complex/render_octomap/render_octomap.cpp:
```diff
@@ -0,0 +1,1167 @@
+	auto descriptorLayout = vkb::initializers::descriptor_set_layout_create_info(setLayoutBindings);
+	VK_CHECK(vkCreateDescriptorSetLayout(get_device().get_handle(), &descriptorLayout, nullptr, &descriptor_set_layout));
+
+	auto pPipelineLayoutCreateInfo = vkb::initializers::pipeline_layout_create_info(&descriptor_set_layout, 1);
```
A little unusual name: `pPipelineLayoutCreateInfo`

Besides that: still a couple of camelCase variables here.

asuessenbach commented on shaders/render_octomap/glsl/imgui.frag:
The imgui shaders aren't used here, anymore?

asuessenbach commented on shaders/render_octomap/glsl/splat.vert:
```diff
@@ -0,0 +1,130 @@
+        outConic    = vec3(0.0);
+        outOpacity  = 0.0;
+        outCoord    = vec2(0.0);
+        return;
```
Could you elaborate a bit how this thing means it is culled?

gpx1000 committed:
Address Andreas' latest comments.

asuessenbach reviewed (approved):
Without visual continuity between the three modes, it's hard to compare them.
But I agree, for this sample showcasing the actual usage, that's probably acceptable.
